### PR TITLE
[Std] [lang] Add ti.rsqrt() and ti.Vector.norm_inv()

### LINKED ISSUE #1288: [Std] [lang] Add ti.rsqrt() and v.norm_rcp()
**Concisely describe the proposed feature**
I would like to add `rsqrt` with optimization for people who care about performance.
For example, GLSL provides `inversesqrt` with better optimization than `1 / sqrt`.

**Describe the solution you'd like (if any)**
In fact the `rsqrt` is already included in `BinaryOpType`, we just need to:
1. Export `taichi_lang_core.expr_rsqrt` using `pybind11`.
1. Add `ti.rsqrt()` to frontend (`python/taichi/lang/ops.py`).
1. Add `ti.Vector.norm_rcp()` (`python/taichi/lang/matrix.py`).
1. Add `BinaryOpType::rsqrt` support to all backends.
1. Do a performance comparison to show the value of your work.

**Additional comments**
See https://en.wikipedia.org/wiki/Fast_inverse_square_root for details about why `rsqrt` is faster than `1 / sqrt`.

### PR DESCRIPTION
**Pull request by archibate**
Related issue = close #1288

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
**I know this will break DiffTaichi if I modify `ti.Vector.normalized()` to use `norm_inv`**, so could you tell me how to make autodiff work for rsqrt?

### CODE REVIEW

**Comment by archibate** [comment]
Running the following test code:
```py
import taichi as ti
import numpy as np
import time
ti.init(ti.opengl)
use_rsqrt = False

N, M = 2048, 256

x = ti.Vector(3, ti.f32, (N, M))

@ti.kernel
def func():
    for i in range(N):
        for j in range(M):
            if ti.static(use_rsqrt):
                x[i, j] = x[i, j] * x[i, j].norm_inv()
            else:
                x[i, j] = x[i, j] / x[i, j].norm()


x.from_numpy(np.random.random((N, M, 3)))
func()
func()
func()

print('starting...')
times = 0.0
for i in range(1000):
    a = time.time()
    func()
    if ti.cfg.arch == ti.cuda:
        ti.sync()
    elif ti.cfg.arch == ti.opengl:
        x.to_numpy()
    times += time.time() - a

print(times, 'ms')
```

obtains:
CPU: `0.88ms -> 0.71ms`.
CUDA: `1.58ms -> 1.57ms`.
Mesa OpenGL: `1.97ms -> 1.97ms`.
NVIDIA OpenGL: `3.01ms -> 3.01ms`.

Well... not sure if it's memory copy overhead or compiler optimization make GPU boost invisible...

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -548,6 +548,11 @@ def norm(self, l=2, eps=0):
         assert l == 2
         return impl.sqrt(self.norm_sqr() + eps)
 
+    @taichi_scope
+    def norm_inv(self, l=2, eps=0):
+        assert l == 2
+        return impl.rsqrt(self.norm_sqr() + eps)
+
     @taichi_scope
     def norm_sqr(self):
```
Comment (yuanming-hu) [inline python/taichi/lang/matrix.py:557]: Probably in a future PR: let's deprecate `norm_sqr` and use `norm_square` instead. Some people will think of `sqr` as *square root*.

**Comment by yuanming-hu** [comment]
> Well... not sure if it's memory copy overhead or compiler optimization make GPU boost invisible...

Your kernel is memory-bound. To make it compute-bound, do something like

```python
@ti.kernel
def func():
    for i in range(N):
        for j in range(M):
            if ti.static(use_rsqrt):
                x[i, j] = x[i, j] * (x[i, j].norm_inv() + x[i, j].norm_inv(eps=1e-1)+ x[i, j].norm_inv(eps=1e-2) + x[i, j].norm_inv(eps=1e-3)+ x[i, j].norm_inv(eps=1e-4))
            else:
                x[i, j] = x[i, j] * (1 / x[i, j].norm() + 1 / x[i, j].norm(eps=1e-1) + 1 / x[i, j].norm(eps=1e-2) + 1 / x[i, j].norm(eps=1e-3))
```

Also note that some backend compilers will do rsqrt for `1 / sqrt` automatically under `fastmath`.

File: python/taichi/lang/matrix.py (hunk 1 above)
Comment (archibate) [reply python/taichi/lang/matrix.py:557]: OK, on my way, but what's the sqrt of norm... may be `sqrt(sqrt(dot(v, v)))`...?

**Comment by archibate** [comment]
Thank for the information! It's `2.0ms -> 1.6ms` on CPU now.

**Comment by archibate** [comment]
Great! Let's merge this for now and add rsqrt autodiff support iapr.

**Pull request merged**